lucene/core/src/java/org/apache/lucene/search/MaxScoreTerminator.java

```diff
@@ -0,0 +1,170 @@
+ * from consideration.</p>
+ *
+ * <p> In practice this leads to a good bound on the number of documents collected, which tend to
+ * exceed numHits by a small factor.  When the documents are evenly distributed among N segments,
+ * we expect to collect approximately (N+1/N) * numHits documents. In a worst case, where *all*
+ * the best documents are in a single segment, we expect to collect something O(log N) ie
+ * (1/N + 1/N-1 + ... + 1) * numHits documents, which is still much better than
+ * the N * numHits we would collect with a naive strategy.
+ * </p>
+ */
+class MaxScoreTerminator {
+
+  // we use 2^5-1 to check the remainder with a bitwise operation
+  // private static final int DEFAULT_INTERVAL_BITS = 10;
+  private static final int DEFAULT_INTERVAL_BITS = 2;
+  private int interval;
+  int intervalMask;
+
+  /** The worst score for each leaf */
+  private final List<LeafState> leafStates;
+
+  /** The total number of docs to collect: from the Collector's numHits */
+  final int totalToCollect;
+
+  /** An upper bound on the number of docs "excluded" from max-score accounting due to early termination. */
+  private int numExcludedBound;
+
+  /** A lower bound on the total hits collected by all leaves */
+  int totalCollected;
+
+  /** the worst hit over all */
+  LeafState leafState;
+
+  MaxScoreTerminator(int totalToCollect) {
+    leafStates = new ArrayList<>();
+    this.totalToCollect = totalToCollect;
+    setIntervalBits(DEFAULT_INTERVAL_BITS);
+  }
+
+  synchronized LeafState add() {
+    LeafState newLeafState = new LeafState();
+    leafStates.add(newLeafState);
+    if (leafState == null) {
+      leafState = newLeafState;
+    }
+    return newLeafState;
+  }
+
+  // for testing
+  void setIntervalBits(int bitCount) {
+    interval = 1 << bitCount;
+    intervalMask = interval - 1;
+  }
+
+  /**
+   * Called by leaf collectors periodically to update their progress.
+   * @param newLeafState the leaf collector's current lowest score
+   * @return whether the collector should terminate
+   */
+  synchronized boolean update(LeafState newLeafState) {
+    totalCollected += interval;
+    //System.out.println(" scoreboard totalCollected = " + totalCollected + "/" + totalToCollect + " "
+    //  + newLeafState + " ? " + leafState + ":" + newLeafState.compareTo(leafState));
+    if (newLeafState.compareTo(leafState) >= 0 || newLeafState.resultCount >= totalToCollect) {
+      leafState = newLeafState;
+      if (totalCollected >= totalToCollect) {
+        // in this case, we have enough scores: remove the worst leaves
+        updateWorstHit();
+        // and tell the current leaf collector to terminate since it can
+        // no longer contribute any top hits
+        return true;
+      }
+    }
+    return false;
+  }
+
+  private void updateWorstHit() {
+    //System.out.println("scoreboard updateWorstHit");
+    if (leafStates.size() > 1) {
+      Collections.sort(leafStates);
```

Comment:
A priority queue should be better here, in general?  It can insert and return the worst `LeafState` in `O(log(N))` time, amortized, whereas this sort is `O(N * log(N))` cost.  But maybe since the number of leaves is typically "smallish", it would not matter much in practice (the constant factors hidden in those big-Os would offset any gains)?

Reply:
Well, also we do not care what the state is in between the relatively infrequent events (I think?) when we cross the `totalToCollect` boundary, so we save the effort of many calls to `updateTop` or similar. In fact I think we can improve this here by only calling `updateWorstHit` when `totalCollected - numExcludedBound >= totalToCollect` rather than the current `totalCollected >= totalToCollect`

Reply:
I suppose we could re-insert all items into a PQ here instead of strictly sorting them though, yeah might be worth it